src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala

```diff
@@ -23,6 +23,7 @@ case class NullConst extends ConstExpr
 case class FuncExpr(function: String, args: List[Expr], isAggregation: Boolean) extends Expr
 case class BinaryOpExpr(lhs: Expr, op: String, rhs: Expr) extends Expr
 case class TypecastExpr(lhs: Expr, rhs: String) extends Expr
+case class Placeholder extends Expr
```

Comment:
Yes, code dealing with body atoms should be modified to handle them as patterns with the ExprPatt case you mention.  It's natural to treat the head atom very differently from body because it does not permit patterns in it and other implicit constraints such as aggregate functions being allowed only in the head.  I think most compilation code already is separate because head is kept as a separate Atom and bodies is now a List[Body] type.  Only the checker seems to be confusing those two to use the same code for head and body, which I think makes it easy to end up in wrong/incomplete checks.

VarPatt is there because variable in a pattern has a subtle difference in meaning than other expressions: it means you want that variable's value to be also bound by the column's value, rather than simply equating the two.  An analogy in C would be the difference between assignments `=` and equality comparison `==`: after both the variable's value should be equal to the other side, but assignment means slightly more than that.  I guess this may not be clear enough with just the simple VarPatt case, but as we introduce pattern matching for complex data types it'll be obvious why we need cases looking similar to Expr's.  To eventually support pattern matching against array/list or protobuf records, e.g., `sentences(id,[t1,_,_,_,t5,restOfTokens...],text)` or `sentences(id,text, nlpTags@{tokens:ts, ners:ns})`, we'll need to add these cases to Patt instead of Expr.

In the compiler, there seems to be already [special cases for VarExpr](https://github.com/HazyResearch/ddlog/blob/a24788e27ac65d4118a518f1b92f9f2d950af02f/src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala#L356-L364), which signals the need for having an explicit type case when we switch to the Patt types.  Otherwise, you'll have to do matches like `case ExprPatt(VarExpr(x)) => ...`.  After all, we're constructing a parse tree instead of keeping everything as flat String to make it clearer and easier to transform the program, hence the more disambiguation you do earlier the simpler your later processing gets.